src/instructlab/feature_gates.py

```diff
@@ -0,0 +1,29 @@
+# Standard
+from enum import StrEnum
+import os
+
+
+class FeatureScopes(StrEnum):
+    Default = "Default"
+    DevPreviewNoUpgrade = "DevPreviewNoUpgrade"
+    TechPreviewNoUpgrade = "TechPreviewNoUpgrade"
+    CustomNoUpgrade = "TechPreviewNoUpgrade"
+
+
+class GatedFeatures(StrEnum):
+    RAG = "RAG"
+
+
+class FeatureGating:
+    env_var_name = "ILAB_FEATURE_SCOPE"
+    _feature_scopes = {GatedFeatures.RAG: FeatureScopes.DevPreviewNoUpgrade}
+
+    @staticmethod
+    def available_scopes():
+        return [f.value for f in FeatureScopes]
+
+    @staticmethod
+    def feature_available(feature: GatedFeatures):
+        return FeatureGating._feature_scopes.get(feature, None) == os.environ.get(
+            FeatureGating.env_var_name, FeatureScopes.Default
+        )
```

Comment:
It's not a big deal for now, since there's only one gated feature, but if we want this feature gating to be generic and usable for multiple features we'll have to consider how someone enables multiple feature gates or whether there's a hierarchy. As a concrete example, should enabling `DevPreviewNoUpgrade` also enable all `TechPreviewNoUpgrade` features? If not, how would I enable both? The same goes for `CustomNoUpgrade` - in effect, is there a hierarchy here where enabling one less-stable gate automatically enables all more stable gates above it? As written today a user can only enable a single level.

Reply:
That's a great question that I think needs to be figured out down the line, but that would be premature to model now. I don't see anything that prevents us from choosing any particular model.

Reply:
If a feature is DevPreview, it is assumed it might really break  your environment. But if a feature is TechPreview it's assumed that what the user is running is a bit closer to production. If someone _just_ wants the new TechPreview features I don't think they should get the DevPreview ones that might break their environment  more than they intended

Reply:
I'm sure that this will come up in the short to medium term future; I don't think this PR is the place to decide it.